autosklearn/ensemble_builder.py

```diff
@@ -567,19 +567,41 @@ def get_n_best_preds(self):
         # One can only read at most max_models_on_disc models
         if self.max_models_on_disc is not None:
             if not isinstance(self.max_models_on_disc, numbers.Integral):
-                worst_consumption = self.get_worst_disk_consumption()
-                if worst_consumption:
-                    num_models = math.floor(self.max_models_on_disc/worst_consumption)
-                    self.logger.debug("Limiting num of models via float max_models_on_disc"
-                                  "worst_consumption={} num_models={}".format(
-                                    worst_consumption,
-                                    num_models
-                                  )
-                                  )
-                    # The first call can be None. From there onwards, always keep
-                    # the minimum number of models
-                    sizes = [self.max_resident_models, int(num_models)]
-                    self.max_resident_models = min( i for i in sizes if i is not None )
+                consumption = [
+                    [
+                        v["ens_score"],
+                        v["disc_space_cost_mb"],
+                     ] for v in self.read_preds.values()
+                ]
+                max_consumption = max(i[1] for i in consumption)
+
+                # We are pessimistic with the consumption limit indicated by
+                # max_models_on_disc by 1 model. Such model is assumed to spend
+                # max_consumption megabytes
+                if (sum(i[1] for i in consumption) + max_consumption) > self.max_models_on_disc:
+
+                    # just leave the best -- higher is better!
+                    # This list is in descending order, to preserve the best models
+                    sorted_cum_consumption = np.cumsum([
+                        i[1] for i in list(reversed(sorted(consumption)))
+                    ])
+                    max_models = np.argmax(sorted_cum_consumption > self.max_models_on_disc)
+
+                    # Make sure that at least 1 model survives
+                    self.max_resident_models = max(1, min(i for i in [
+                        self.max_resident_models,  # Can be None in first iter
```

Comment:
Do we have to compare about the previous max_resident_models? It might be possible that we increase the number of models we have on disc if the good models (which are found later) use less memory than bad models, right?

Reply:
The original motivation to do this was to have a static number so that we do not run into the scenario that we need models that are deleted (Like what happens with the float mem in disk in previous versions).

I have been reading the code for a while and doing dummy scenarios, and I cannot think of a scenario that justifies not increasing max resident models (because score predictions always regenerate the list of files and annotate the prediction file size).

I will move max resident models to the new value regardless of the past value.

Reply:
I don't see an issue here either and think the new code is fine.